streamtrace.cc

```diff
@@ -1695,6 +1695,51 @@ void decode_cap(capability_register &cap, uint64_t val2, uint64_t val3,
 	cap.length = val5;
 	cap.offset = val3 - cap.base;
 }
+/**
+ * Decode a 128-bit capability register from streamtrace values.  `val2` is always
+ * `val2` from the streamtrace, but `val1` is either `val1` or `pc` depending
+ * on the trace format. val1/pc is the 1st half of the capability containing the
+ * pointer and val2 is the 2nd half of the compressed capability containing the
+ * bounds and permissions. 
+ */
+void decode_cap128(capability_register &cap, uint64_t val2, uint64_t val1)
+{
+	// extract the tag from the reserved bits where it has been stashed for tracing.
+	cap->valid = extract_bits<47>(val2);
+	cap->unsealed =	extract_bits<40>(val2);
+	cap->type =	0; // TODO extract type from typed capabilities
+	cap->permissions = extract_bits<63,49>(val2);
+	uint64_t exp     = extract_bits<46,41>(val2);
+	uint64_t mask    = ((uint64_t) ~0)<<20ULL;
```

Comment:
I think I can figure out what this is doing, but please add a comment explaining it.